src/Formik.tsx

```diff
@@ -1,306 +1,240 @@
+  | { type: 'SET_STATUS'; payload: any }
+  | { type: 'SET_FORMIK_STATE'; payload: FormikState<Values> }
+  | { type: 'RESET_FORM'; payload: FormikState<Values> };
+
+// State reducer
+function formikReducer<Values>(
+  state: FormikState<Values>,
+  msg: FormikMessage<Values>
+) {
+  switch (msg.type) {
+    case 'SET_VALUES':
+      return { ...state, values: msg.payload };
+    case 'SET_TOUCHED':
+      return { ...state, touched: msg.payload };
+    case 'SET_ERRORS':
+      return { ...state, errors: msg.payload };
+    case 'SET_STATUS':
+      return { ...state, status: msg.payload };
+    case 'SET_ISSUBMITTING':
+      return { ...state, isSubmitting: msg.payload };
+    case 'SET_FIELD_VALUE':
+      return {
+        ...state,
+        values: setIn(state.values, msg.payload.field, msg.payload.value),
+      };
+    case 'SET_FIELD_TOUCHED':
+      return {
+        ...state,
+        touched: setIn(state.touched, msg.payload.field, msg.payload.value),
+      };
+    case 'SET_FIELD_ERROR':
+      return {
+        ...state,
+        errors: setIn(state.errors, msg.payload.field, msg.payload.value),
+      };
+    case 'RESET_FORM':
+    case 'SET_FORMIK_STATE':
+      return { ...state, ...msg.payload };
+    case 'SUBMIT_ATTEMPT':
+      return {
+        ...state,
+        touched: setNestedObjectValues<FormikTouched<Values>>(
+          state.values,
+          true
+        ),
+        isSubmitting: true,
+        isValidating: true,
+        submitCount: state.submitCount + 1,
+      };
+    case 'SUBMIT_FAILURE':
+      return {
+        ...state,
+        isSubmitting: false,
+      };
+    case 'SUBMIT_SUCCESS':
+      return {
+        ...state,
+        isSubmitting: false,
+      };
+    default:
+      return state;
   }
+}
 
-  componentWillUnmount() {
-    // This allows us to prevent setting state on an
-    // unmounted component. This can occur if Formik is in a modal, and submission
-    // toggles show/hide, and validation of a blur field takes longer than validation
-    // before a submit.
-    // @see https://github.com/jaredpalmer/formik/issues/597
-    // @see https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html
-    this.didMount = false;
-
-    // Cancel validation on unmount.
-    if (this.validator) {
-      this.validator();
-    }
-  }
+export function useFormik<Values = object>({
+  validateOnChange = true,
+  validateOnBlur = true,
+  isInitialValid = false,
+  debounceValidationMs = 300,
+  ...rest
+}: FormikConfig<Values>) {
+  const props = { validateOnChange, validateOnBlur, isInitialValid, ...rest };
+  const initialValues = React.useRef(props.initialValues);
+  const didMount = React.useRef<boolean>(false);
+  const fields = React.useRef<{
+    [field: string]: {
+      validate: (value: any) => string | Promise<string> | undefined;
+    };
+  }>({});
+  React.useEffect(
+    () => {
+      initialValues.current = props.initialValues;
+    },
+    [props.initialValues]
```

Comment:
What this for? Why not just assign in render without effect?

Reply:
Yeah, you are right

Reply:
Fixed.